Sleep timer crashes on Android 14

**Have you read the [FAQ](https://github.com/RetroMusicPlayer/RetroMusicPlayer/blob/master/FAQ.md)?**
[Yes]

**Has the bug already been reported? Please search in GitHub issue tab before creating an issue.**
[No]

**Describe the bug**
Sleep timer functionality does not work on Android 14 anymore

**How To Reproduce**
Steps to reproduce the behavior:
1.  Install the app on an Android 14 device
2. Play any song and start the sleep timer
3. Click the start button
4. App Crash screen appears


**Expected behavior**
Sleep timer works as expected

**Crash log**
If the app is crashing, add a crash log
<details>
  <summary>Click to view logs</summary>
Build version: 6.1.0 
Current date: 2023-12-27 19:56:40 
Device: Google Pixel 6a 
OS version: Android 14 (SDK 34) 
 
Stack trace:  
java.lang.SecurityException: Caller code.name.monkey.retromusic needs to hold android.permission.SCHEDULE_EXACT_ALARM or android.permission.USE_EXACT_ALARM to set exact alarms.
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
	at android.os.Parcel.createException(Parcel.java:3041)
	at android.os.Parcel.readException(Parcel.java:3024)
	at android.os.Parcel.readException(Parcel.java:2966)
	at android.app.IAlarmManager$Stub$Proxy.set(IAlarmManager.java:311)
	at android.app.AlarmManager.setImpl(AlarmManager.java:1084)
	at android.app.AlarmManager.setImpl(AlarmManager.java:1044)
	at android.app.AlarmManager.setExact(AlarmManager.java:808)
	at e3.j.onClick(R8$$SyntheticClass:98)
	at androidx.appcompat.app.AlertController$c.handleMessage(AlertController.java:38)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8194)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.alarm.AlarmManagerService$5.set(AlarmManagerService.java:2981)
	at android.app.IAlarmManager$Stub.onTransact(IAlarmManager.java:181)
	at android.os.Binder.execTransactInternal(Binder.java:1332)
	at android.os.Binder.execTransact(Binder.java:1278)


</details>

**Device info:**
 - Device: Pixel 6A
 - Android version: Android 14
 - App version 6.1.0

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -65,7 +65,7 @@ android {
     }
     lint {
         abortOnError true
-        warning 'ImpliedQuantity', 'Instantiatable', 'MissingQuantity', 'MissingTranslation'
+        warning 'ImpliedQuantity', 'Instantiatable', 'MissingQuantity', 'MissingTranslation', 'StringFormatInvalid'
     }
     compileOptions {
         sourceCompatibility JavaVersion.VERSION_17

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -6,7 +6,9 @@
         android:name="android.permission.BLUETOOTH"
         android:maxSdkVersion="30" />
     <uses-permission android:name="android.permission.DISABLE_KEYGUARD" />
-    <uses-permission android:name="android.permission.SCHEDULE_EXACT_ALARM" />
+    <uses-permission
+        android:name="android.permission.SCHEDULE_EXACT_ALARM"
+        tools:ignore="ProtectedPermissions" />
     <uses-permission android:name="android.permission.VIBRATE" />
     <uses-permission android:name="android.permission.WAKE_LOCK" />
     <uses-permission

--- a/app/src/main/java/code/name/monkey/retromusic/activities/PermissionActivity.kt
+++ b/app/src/main/java/code/name/monkey/retromusic/activities/PermissionActivity.kt
@@ -15,6 +15,7 @@
 package code.name.monkey.retromusic.activities
 
 import android.Manifest.permission.BLUETOOTH_CONNECT
+import android.app.AlarmManager
 import android.content.Intent
 import android.content.pm.PackageManager
 import android.content.res.ColorStateList
@@ -24,6 +25,7 @@ import android.provider.Settings
 import androidx.activity.OnBackPressedCallback
 import androidx.annotation.RequiresApi
 import androidx.core.app.ActivityCompat
+import androidx.core.content.getSystemService
 import androidx.core.net.toUri
 import androidx.core.text.parseAsHtml
 import androidx.core.view.isVisible
@@ -66,6 +68,11 @@ class PermissionActivity : AbsMusicServiceActivity() {
                     arrayOf(BLUETOOTH_CONNECT),
                     BLUETOOTH_PERMISSION_REQUEST
                 )
+            }
+            binding.alarmPermission.show()
+            binding.alarmPermission.setButtonClick {
+                val intent = Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM)
+                startActivity(intent)
             }
         } else {
             binding.audioPermission.setNumber("2")
@@ -124,6 +131,11 @@ class PermissionActivity : AbsMusicServiceActivity() {
                 binding.bluetoothPermission.checkImage.imageTintList =
                     ColorStateList.valueOf(accentColor())
             }
+            if (hasAlarmPermission()) {
+                binding.alarmPermission.checkImage.isVisible = true
+                binding.alarmPermission.checkImage.imageTintList =
+                    ColorStateList.valueOf(accentColor())
+            }
         }
     }
 
@@ -142,5 +154,10 @@ class PermissionActivity : AbsMusicServiceActivity() {
     @RequiresApi(Build.VERSION_CODES.M)
     private fun hasAudioPermission(): Boolean {
         return Settings.System.canWrite(this)
+    }
+
+    @RequiresApi(Build.VERSION_CODES.S)
+    private fun hasAlarmPermission(): Boolean {
+        return getSystemService<AlarmManager>()?.canScheduleExactAlarms() == true
     }
 }

--- a/app/src/main/java/code/name/monkey/retromusic/dialogs/SleepTimerDialog.kt
+++ b/app/src/main/java/code/name/monkey/retromusic/dialogs/SleepTimerDialog.kt
@@ -19,9 +19,11 @@ import android.app.Dialog
 import android.app.PendingIntent
 import android.content.DialogInterface
 import android.content.Intent
+import android.os.Build
 import android.os.Bundle
 import android.os.CountDownTimer
 import android.os.SystemClock
+import android.provider.Settings
 import android.widget.CheckBox
 import android.widget.SeekBar
 import android.widget.TextView
@@ -129,17 +131,27 @@ class SleepTimerDialog : DialogFragment() {
                         SystemClock.elapsedRealtime() + minutes * 60 * 1000
                     PreferenceUtil.nextSleepTimerElapsedRealTime = nextSleepTimerElapsedTime.toInt()
                     val am = requireContext().getSystemService<AlarmManager>()
-                    am?.setExact(
-                        AlarmManager.ELAPSED_REALTIME_WAKEUP,
-                        nextSleepTimerElapsedTime,
-                        pi
-                    )
 
-                    Toast.makeText(
-                        requireContext(),
-                        requireContext().resources.getString(R.string.sleep_timer_set, minutes),
-                        Toast.LENGTH_SHORT
-                    ).show()
+                    if (VersionUtils.hasS() && am?.canScheduleExactAlarms() != true) {
+                        Toast.makeText(
+                            requireContext(),
+                            requireContext().resources.getString(R.string.sleep_timer_no_permission),
+                            Toast.LENGTH_SHORT
+                        ).show()
+                        val intent = Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM)
+                        startActivity(intent)
+                    } else {
+                        am?.setExact(
+                           AlarmManager.ELAPSED_REALTIME_WAKEUP,
+                           nextSleepTimerElapsedTime,
+                           pi
+                       )
+                        Toast.makeText(
+                            requireContext(),
+                            requireContext().resources.getString(R.string.sleep_timer_set, minutes),
+                            Toast.LENGTH_SHORT
+                        ).show()
+                    }
                 }
             }
             setView(binding.root)

--- a/app/src/main/res/layout/activity_permission.xml
+++ b/app/src/main/res/layout/activity_permission.xml
@@ -78,6 +78,19 @@
                 app:permissionTitleNumber="3"
                 app:permissionTitleSubTitle="@string/ringtone_summary"
                 tools:visibility="visible" />
+
+            <code.name.monkey.retromusic.views.PermissionItem
+                android:id="@+id/alarm_permission"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_weight="@integer/permission_layout_weight"
+                android:visibility="gone"
+                app:permissionButtonTitle="@string/grant_access"
+                app:permissionIcon="@drawable/ic_phonelink_ring"
+                app:permissionTitle="@string/permission_alarm_title"
+                app:permissionTitleNumber="4"
+                app:permissionTitleSubTitle="@string/permission_alarm_summary"
+                tools:visibility="visible" />
         </LinearLayout>
     </ScrollView>
 

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -456,6 +456,8 @@
     <string name="retro_music_pro">Retro Music Pro</string>
     <string name="ringtone_summary">The app needs permission to access your device settings in order to set music as Ringtone</string>
     <string name="ringtone_title">Ringtone (Optional)</string>
+    <string name="permission_alarm_summary">The app needs permission to schedule exact alarms to set sleep timers</string>
+    <string name="permission_alarm_title">Alarm (Optional)</string>
     <string name="saf_delete_failed">File delete failed: %s</string>
     <!-- SAF -->
     <string name="saf_error_uri">Can\'t get SAF URI</string>
@@ -563,4 +565,5 @@
     <string name="you_will_be_forwarded_to_the_issue_tracker_website">You will be forwarded to the issue tracker website.</string>
     <string name="your_account_data_is_only_used_for_authentication">Your account data is only used for authentication.</string>
     <string name="could_not_write_tags_to_file">Could not write tags to the music file!</string>
+    <string name="sleep_timer_no_permission">No permission to schedule alarms, please allow Retro Music.</string>
 </resources>

